File: internal/domain/download/segment.go
package download

import "time"

// SegmentID identifies a Segment. Allocated by the persistence layer.
type SegmentID int64

// Segment is one Usenet article — the smallest unit of work for the
// download orchestrator. Each segment maps to a Message-ID, a byte
// range within the parent file, and a per-attempt retry count.
//
// Segments are entities owned by their File, not aggregates. Mutation
// happens through methods on the parent Job (which owns the File).
type Segment struct {
	id        SegmentID
	fileID    FileID
	seqIndex  int    // 1-based within the parent file
	messageID string // without surrounding angle brackets
	bytes     int64  // article size on the wire (overhead included)
	state     SegmentState
	attempts  int
	lastError string
	// fileOffset is the 0-based byte offset within the assembled file
	// where this segment's decoded bytes go. Set after the first
	// successful yEnc decode (from =ypart begin - 1) and persisted.
	fileOffset int64
	// nextRetryAt is the earliest time this segment is eligible for
	// re-dispatch. Zero = ready now. Populated on transient retry so
	// the back-off survives a restart instead of being lost with the
	// in-memory goroutine that was sleeping on time.After.
	nextRetryAt time.Time
}

// NewSegmentParams constructs a fresh pending Segment.
type NewSegmentParams struct {
	SeqIndex  int
	MessageID string
	Bytes     int64
}

func newSegment(p NewSegmentParams) *Segment {
	return &Segment{
		seqIndex:  p.SeqIndex,
		messageID: p.MessageID,
		bytes:     p.Bytes,
		state:     SegmentStatePending,
	}
}

// HydrateSegmentParams is what the repository hands back when loading
// a row.
type HydrateSegmentParams struct {
	ID          SegmentID
	FileID      FileID
	SeqIndex    int
	MessageID   string
	Bytes       int64
	State       SegmentState
	Attempts    int
	LastError   string
	FileOffset  int64
	NextRetryAt time.Time
}

// HydrateSegment is the adapter-side constructor that reconstructs a
// Segment from persistence. No events are emitted.
func HydrateSegment(p HydrateSegmentParams) *Segment {
	return &Segment{
		id:          p.ID,
		fileID:      p.FileID,
		seqIndex:    p.SeqIndex,
		messageID:   p.MessageID,
		bytes:       p.Bytes,
		state:       p.State,
		attempts:    p.Attempts,
		lastError:   p.LastError,
		fileOffset:  p.FileOffset,
		nextRetryAt: p.NextRetryAt,
	}
}

// Accessors.
func (s *Segment) ID() SegmentID       { return s.id }
func (s *Segment) FileID() FileID      { return s.fileID }
func (s *Segment) SeqIndex() int       { return s.seqIndex }
func (s *Segment) MessageID() string   { return s.messageID }
func (s *Segment) Bytes() int64        { return s.bytes }
func (s *Segment) State() SegmentState { return s.state }
func (s *Segment) Attempts() int       { return s.attempts }
func (s *Segment) LastError() string   { return s.lastError }
func (s *Segment) FileOffset() int64   { return s.fileOffset }

// NextRetryAt returns the earliest instant this segment is eligible
// for re-dispatch. Zero = ready now.
func (s *Segment) NextRetryAt() time.Time { return s.nextRetryAt }

// MarkPendingRetry flips a transient-failed (inflight/failed) segment
// back to pending with a durable backoff. After saving, the
// orchestrator's pending-segment query (filtered by next_retry_at)
// will skip this segment until `at` passes — even across a process
// restart.
//
// Records the error string so the per-job timeline + UI surface what
// went wrong on the previous attempt without callers having to track
// it separately.
func (s *Segment) MarkPendingRetry(at time.Time, errMsg string) {
	s.state = SegmentStatePending
	s.nextRetryAt = at.UTC()
	s.attempts++
	s.lastError = errMsg
}

// SetID is called by the repository to assign a database id after
// insert.
func (s *Segment) SetID(id SegmentID) { s.id = id }

// SetFileID is called when a fresh segment is associated with its
// (newly-saved) file row.
func (s *Segment) SetFileID(id FileID) { s.fileID = id }

// SetFileOffset records the 0-based byte offset within the assembled
// file. The orchestrator computes this after decoding the first
// segment of a file (single-part: 0; multi-part: ypart.begin - 1).
func (s *Segment) SetFileOffset(off int64) { s.fileOffset = off }

File: src/domain/download/segment.zig
//! `Segment` — one Usenet article, the smallest unit of work for the
//! download orchestrator.
//!
//! Segments are entities owned by their `File`, which is owned by the
//! `Job` aggregate. They hold no allocator: every string in here is
//! allocated and freed by the owning Job (see `job.zig`), which is what
//! keeps ownership auditable — there is exactly one `deinit` in this
//! whole context.
//!
//! Mutation is likewise the Job's business. The methods below are the
//! ones that cannot break an aggregate invariant (id plumbing after an
//! INSERT, offset bookkeeping); anything that changes `state` lives on
//! `Job` because it has to keep the unresolved-segment counters in
//! step.

const std = @import("std");
const state_mod = @import("state.zig");

const SegmentId = state_mod.SegmentId;
const FileId = state_mod.FileId;
const SegmentState = state_mod.SegmentState;
const Timestamp = state_mod.Timestamp;

/// Constructor input for a fresh, pending Segment. The strings are
/// borrowed for the duration of the call; the Job copies what it keeps.
pub const NewSegmentParams = struct {
    seq_index: i32 = 0,
    message_id: []const u8,
    bytes: i64 = 0,
};

/// Constructor input for a Segment being rebuilt from persistence.
/// Strings are borrowed for the duration of the call.
pub const HydrateSegmentParams = struct {
    id: SegmentId = 0,
    file_id: FileId = 0,
    seq_index: i32 = 0,
    message_id: []const u8,
    bytes: i64 = 0,
    state: SegmentState = .pending,
    attempts: i32 = 0,
    last_error: ?[]const u8 = null,
    file_offset: i64 = 0,
    next_retry_at: ?Timestamp = null,
};

pub const Segment = struct {
    /// Assigned by the persistence layer on INSERT; 0 until then.
    id: SegmentId = 0,
    file_id: FileId = 0,
    /// 1-based position within the parent file.
    seq_index: i32 = 0,
    /// Message-id without the surrounding angle brackets. Owned by the
    /// Job; immutable once constructed, which is why events may borrow
    /// it instead of copying.
    message_id: []const u8,
    /// Article size on the wire, yEnc overhead included.
    bytes: i64 = 0,
    /// Read-only from outside the aggregate: transition it through the
    /// `Job.markSegment*` methods so the resolved counters stay honest.
    state: SegmentState = .pending,
    attempts: i32 = 0,
    /// Last failure text, or null for "no error recorded". Owned by the
    /// Job, which frees the previous value before storing a new one.
    last_error: ?[]const u8 = null,
    /// 0-based byte offset within the assembled file where this
    /// segment's decoded bytes go. Set after the first successful yEnc
    /// decode (`=ypart begin - 1`) and persisted.
    file_offset: i64 = 0,
    /// Earliest instant this segment may be re-dispatched; null = ready
    /// now. Populated on transient retry so the back-off survives a
    /// restart instead of dying with an in-memory timer.
    next_retry_at: ?Timestamp = null,

    /// Go returned "" for "no error"; callers that want the same shape
    /// use this instead of unwrapping the optional.
    pub fn lastError(self: Segment) []const u8 {
        return self.last_error orelse "";
    }

    /// Called by the repository to assign a database id after INSERT.
    /// The Job's segment index keys on the id, so a caller doing this
    /// must follow up with `Job.rebuildSegmentIndex`.
    pub fn setId(self: *Segment, id: SegmentId) void {
        self.id = id;
    }

    /// Called when a fresh segment is associated with its newly-saved
    /// file row.
    pub fn setFileId(self: *Segment, id: FileId) void {
        self.file_id = id;
    }

    /// Records the 0-based byte offset within the assembled file. The
    /// orchestrator computes this after decoding the first segment of a
    /// file (single-part: 0; multi-part: `ypart.begin - 1`).
    pub fn setFileOffset(self: *Segment, off: i64) void {
        self.file_offset = off;
    }

    /// True when the segment is eligible for dispatch at `now`: pending
    /// and past any back-off window. Shared by `Job.pendingSegments`
    /// and `Job.nextRetryReadyAt` so the two can never disagree about
    /// what "ready" means.
    pub fn isReadyAt(self: Segment, now: Timestamp) bool {
        if (self.state != .pending) return false;
        const at = self.next_retry_at orelse return true;
        return at <= now;
    }

    /// True when the segment is pending but held back by its back-off
    /// window — the state `nextRetryReadyAt` reports on.
    pub fn isDeferredAt(self: Segment, now: Timestamp) bool {
        if (self.state != .pending) return false;
        const at = self.next_retry_at orelse return false;
        return at > now;
    }
};
